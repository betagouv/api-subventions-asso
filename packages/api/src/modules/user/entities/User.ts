import { ObjectId, WithId } from "mongodb";

export default class User {
    constructor(
        public email: string,
        public hashPassword: string,
        public roles: string[],
        public jwt: { token: string, expirateDate: Date },
        public active: boolean,
        public stats: {
            searchCount: number
        },
        public _id?: ObjectId
    ){}
}

export type UserWithoutSecret = Omit<Omit<WithId<User>, 'hashPassword'>, "jwt">;