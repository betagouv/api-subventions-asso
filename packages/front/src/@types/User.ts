export default interface User {
    token: string;
}
