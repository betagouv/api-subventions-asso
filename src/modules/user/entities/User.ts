
export default class User {
    constructor(
        public email: string,
        public hashPassword: string,
        public roles: string[],
        public jwt: { token: string, expirateDate: Date },
        public active: boolean
    ){}
}

export type UserWithoutSecret = Omit<Omit<User, 'hashPassword'>, "jwt">;