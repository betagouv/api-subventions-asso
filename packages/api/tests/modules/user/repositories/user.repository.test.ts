import { ObjectId } from "mongodb";
import User, { UserWithoutSecret } from "../../../../src/modules/user/entities/User";
import userRepository from "../../../../src/modules/user/repositories/user.repository";
import userService from "../../../../src/modules/user/user.service"

describe("UserRepository", () => {

    const defaultUser = {
        email: "test22@beta.gouv.fr",
        hashPassword: "PASSWORD",
        roles: ["user"],
        signupAt: new Date(),
        jwt: { token: "TOKEN", expirateDate: new Date() },
        active: false,
        stats: { searchCount: 0, lastSearchDate: null }
    };
    describe('The methods must not return any secret', () => {
        beforeEach(async () => {
            await userService.createUser("test@beta.gouv.fr");
        });

        it("findByEmail", async () => {
            await expect(userRepository.findByEmail("test@beta.gouv.fr"))
                .resolves
                .toMatchObject(
                    expect.not.objectContaining({ hashPassword: expect.any(String), jwt: { token: expect.any(String), expirateDate: expect.any(Date) } })
                )
        });

        it("update", async () => {
            const user = await userRepository.findByEmail("test@beta.gouv.fr") as UserWithoutSecret;
            await expect(userRepository.update({ ...user, active: true }))
                .resolves
                .toMatchObject(
                    expect.not.objectContaining({ hashPassword: expect.any(String), jwt: { token: expect.any(String), expirateDate: expect.any(Date) } })
                )
        });

        it("create", async () => {
            await expect(userRepository.create(new User(defaultUser)))
                .resolves
                .toMatchObject(
                    expect.not.objectContaining({ hashPassword: expect.any(String), jwt: { token: expect.any(String), expirateDate: expect.any(Date) } })
                )
        });
    });

    describe('removeSecrets', () => {
        it("should remove all secret in user", () => {
            const user = new User(defaultUser);

            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            expect(userRepository.removeSecrets(user)).not.toHaveProperty("jwt")

            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            expect(userRepository.removeSecrets(user)).not.toHaveProperty("hashPassword")
        })
    });

    describe("findPassword", () => {
        beforeEach(async () => {
            await userService.createUser("test@beta.gouv.fr");
        });

        it("should return password", async () => {
            const user = await userRepository.findByEmail("test@beta.gouv.fr") as UserWithoutSecret;

            await expect(typeof await userRepository.findPassword(user)).toBe("string");
        })

        it("should return null", async () => {
            const user = new User(defaultUser);

            await expect(userRepository.findPassword(user as UserWithoutSecret)).resolves.toBe(null);
        })
    })

    describe("findJwt", () => {
        beforeEach(async () => {
            await userService.createUser("test@beta.gouv.fr");
        });

        it("should return jwt", async () => {
            const user = await userRepository.findByEmail("test@beta.gouv.fr") as UserWithoutSecret;

            await expect(userRepository.findJwt(user)).resolves.toMatchObject({ token: expect.any(String), expirateDate: expect.any(Date) });
        })

        it("should return null", async () => {
            const user = new User(defaultUser, new ObjectId());

            await expect(userRepository.findJwt(user as UserWithoutSecret)).resolves.toBe(null);
        })
    })
})