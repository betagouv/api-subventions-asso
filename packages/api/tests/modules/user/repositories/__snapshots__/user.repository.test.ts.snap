// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRepository getUserWithSecretsByEmail should return user 1`] = `
Object {
  "_id": Any<ObjectId>,
  "active": false,
  "email": "test@beta.gouv.fr",
  "firstName": null,
  "hashPassword": Any<String>,
  "jwt": Object {
    "expirateDate": Any<Date>,
    "token": Any<String>,
  },
  "lastName": null,
  "profileToComplete": true,
  "roles": Array [
    "user",
  ],
  "signupAt": Any<Date>,
}
`;
